data/lua/mmbn3_connector.lua

```diff
@@ -0,0 +1,1065 @@
+    ["[ModTools2]"]=0x69,["[ModTools3]"]=0x6A,['Σ']=0x6B,['Ω']=0x6C,['α']=0x6D,['β']=0x6E,['#']=0x6F,['…']=0x70,['>']=0x71,
+    ['<']=0x72,['エ']=0x73,["[BowneGlobal1]"]=0x74,["[BowneGlobal2]"]=0x75,["[BowneGlobal3]"]=0x76,["[BowneGlobal4]"]=0x77,
+    ["[BowneGlobal5]"]=0x78,["[BowneGlobal6]"]=0x79,["[BowneGlobal7]"]=0x7A,["[BowneGlobal8]"]=0x7B,["[BowneGlobal9]"]=0x7C,
+    ["[BowneGlobal10]"]=0x7D,["[BowneGlobal11]"]=0x7E,['\n']=0xE8
+}
+
+local TableConcat = function(t1,t2)
+   for i=1,#t2 do
+      t1[#t1+1] = t2[i]
+   end
+   return t1
+end
+local int32ToByteList_le = function(x)
+    bytes = {}
+    hexString = string.format("%08x", x)
+    for i=#hexString, 1, -2 do
+      hbyte = hexString:sub(i-1, i)
+      table.insert(bytes,tonumber(hbyte,16))
+    end
+    return bytes
+end
+local int16ToByteList_le = function(x)
+    bytes = {}
+    hexString = string.format("%04x", x)
+    for i=#hexString, 1, -2 do
+      hbyte = hexString:sub(i-1, i)
+      table.insert(bytes,tonumber(hbyte,16))
+    end
+    return bytes
+end
+
+local IsInMenu = function()
+    return bit.band(memory.read_u8(0x0200027A),0x10) ~= 0
+end
+local IsInTransition = function()
+    return bit.band(memory.read_u8(0x02001880), 0x10) ~= 0
+end
+local IsInDialog = function()
+    return bit.band(memory.read_u8(0x02009480),0x01) ~= 0
+end
+local IsInBattle = function()
+    return memory.read_u8(0x020097F8) == 0x08
+end
+local IsItemQueued = function()
+    return memory.read_u8(0x2000224) == 0x01
+end
+
+-- This function actually determines when you're on ANY full-screen menu (navi cust, link battle, etc.) but we
+-- don't want to check any locations there either so it's fine.
+local IsOnTitle = function()
+    return bit.band(memory.read_u8(0x020097F8),0x04) == 0
+end
+local IsItemable = function()
+    return not IsInMenu() and not IsInTransition() and not IsInDialog() and not IsInBattle() and not IsOnTitle() and not IsItemQueued()
+end
+
+local is_game_complete = function()
+    if IsOnTitle() or itemState == ITEMSTATE_NONINITIALIZED then return game_complete end
+
+    -- If the game is already marked complete, do not read memory
+    if game_complete then return true end
+    local is_alpha_defeated = bit.band(memory.read_u8(0x2000433), 0x01) ~= 0
+
+    if (is_alpha_defeated) then
+        game_complete = true
+        return true
+    end
+
+    -- Game is still ongoing
+    return false
+end
+
+local saveItemIndexToRAM = function(newIndex)
+    memory.write_s16_le(0x20000AE,newIndex)
+end
+
+local loadItemIndexFromRAM = function()
+    last_index = memory.read_s16_le(0x20000AE)
+    if (last_index < 0) then
+        last_index = 0
+        saveItemIndexToRAM(0)
+    end
+    return last_index
+end
+
+local loadPlayerNameFromROM = function()
+    return memory.read_bytes_as_array(0x7FFFC0,63,"ROM")
+end
+
+-- TODO pass the whole chunk of data and check locations on the other end
```

Comment:
Would highly recommend doing this if it's not too much trouble (I don't know exactly how these flags are laid out, but they look fairly compact). Lots of duplicated information and multiple reads of the same byte below. And if this is ever phased out in favor of SNI, you'll already have that code in python.

Reply:
Awesome, looks much nicer.